### ISSUE TITLE
Randomize features enablement order in tests

### ISSUE THREAD

**Issue by kbr-scylla** [scylladb/scylladb#16014]
Weird and hard to detect bugs happen when correctness depends on the order of enabling features.
See https://github.com/scylladb/scylladb/issues/16004.

We should randomize the order, perhaps in debug mode only, during testing, to quickly catch such dependencies and eliminate them.

cc @denesb @piodul @tgrabiec @avikivity

**Comment by kostja** [comment]
I would move in a different direction: we should enable features in the same order, all the time. We're exercising and testing code paths that have nothing to do with what we want to deliver to the customers. We never change the order of features we add, newer features are added in newer releases. So it's best to enable features in git history order, and test only that.

**Comment by denesb** [comment]
> I would move in a different direction: we should enable features in the same order, all the time. We're exercising and testing code paths that have nothing to do with what we want to deliver to the customers. We never change the order of features we add, newer features are added in newer releases. So it's best to enable features in git history order, and test only that.

Features are stored in an un-ordered container. Furthermore, nothing stops developers from adding features into the middle of the member list. Guaranteeing an ordering of features would be hard and would require ongoing discipline and attention when modifying this code. Of course a simple test could be added to verify the order.

**Comment by kostja** [comment]
> > I would move in a different direction: we should enable features in the same order, all the time. We're exercising and testing code paths that have nothing to do with what we want to deliver to the customers. We never change the order of features we add, newer features are added in newer releases. So it's best to enable features in git history order, and test only that.
> 
> Features are stored in an un-ordered container. Furthermore, nothing stops developers from adding features into the middle of the member list. Guaranteeing an ordering of features would be hard and would require ongoing discipline and attention when modifying this code. Of course a simple test could be added to verify the order.

Well, that's exactly what I am saying. We need to stop doing that. It creates unnecessary complexity out of nowhere. We can change the container to an ordered one. We can easily enforce the feature addition discipline by assigning each feature a unique monotonic numeric ID just like we do with RPC and not allowing to change it in the future.

**Comment by kostja** [comment]
Right, and my point is that we don't need to ever "shake out" the second. Let them interdepend, we don't ever need to fix it. If we implement randomization we will have to address the design issue that is not relevant to any customer-facing product feature.

**Comment by denesb** [comment]
The problem is that when something working correctly depends on timing... all you need is a fluke of slow-down for things to break for apparently no reason. Better have this in tests, than at customer deployments.

**Comment by kostja** [comment]
It will not depend on timing if enabling feature X will not only wait for X in the entire cluster, but that X-1 is already enabled.

**Comment by denesb** [comment]
> It will not depend on timing if enabling feature X will not only wait for X in the entire cluster, but that X-1 is already enabled.

Right, that could work. I don't have a strong opinion on this, I'm all for the simplest possible solution, that will guarantee (as much as possible) that we won't run into bugs like #16004 in the future.
